lib/Dialect/FIRRTL/FIRRTLFieldSource.cpp

```diff
@@ -26,119 +26,74 @@
-    return visitInst(inst);
-  if (auto inst = dyn_cast<InstanceChoiceOp>(op))
-    return visitInstChoice(inst);
-
-  // Track all other definitions of aggregates.
-  if (op->getNumResults()) {
-    auto type = op->getResult(0).getType();
-    if (auto baseType = type_dyn_cast<FIRRTLBaseType>(type);
-        baseType && !baseType.isGround())
-      makeNodeForValue(op->getResult(0), op->getResult(0), {},
-                       foldFlow(op->getResult(0)));
+template <typename Fn>
+FieldSource::PathNode FieldSource::visitValue(Value v, Fn update) const {
+  // Blockargs are never cached, just regenerate.
+  if (auto arg = dyn_cast<BlockArgument>(v)) {
+    return PathNode{v, 0};
   }
-}
-
-void FieldSource::visitSubfield(SubfieldOp sf) {
-  auto value = sf.getInput();
-  const auto *node = nodeForValue(value);
-  assert(node && "node should be in the map");
-  auto sv = node->path;
-  sv.push_back(sf.getFieldIndex());
-  makeNodeForValue(sf.getResult(), node->src, sv, foldFlow(sf));
-}
-
-void FieldSource::visitOpenSubfield(OpenSubfieldOp sf) {
-  auto value = sf.getInput();
-  const auto *node = nodeForValue(value);
-  assert(node && "node should be in the map");
-  auto sv = node->path;
-  sv.push_back(sf.getFieldIndex());
-  makeNodeForValue(sf.getResult(), node->src, sv, foldFlow(sf));
-}
-
-void FieldSource::visitSubindex(SubindexOp si) {
-  auto value = si.getInput();
-  const auto *node = nodeForValue(value);
-  assert(node && "node should be in the map");
-  auto sv = node->path;
-  sv.push_back(si.getIndex());
-  makeNodeForValue(si.getResult(), node->src, sv, foldFlow(si));
-}
-
-void FieldSource::visitOpenSubindex(OpenSubindexOp si) {
-  auto value = si.getInput();
-  const auto *node = nodeForValue(value);
-  assert(node && "node should be in the map");
-  auto sv = node->path;
-  sv.push_back(si.getIndex());
-  makeNodeForValue(si.getResult(), node->src, sv, foldFlow(si));
-}
-
-void FieldSource::visitSubaccess(SubaccessOp sa) {
-  auto value = sa.getInput();
-  const auto *node = nodeForValue(value);
-  assert(node && "node should be in the map");
-  auto sv = node->path;
-  sv.push_back(-1);
-  makeNodeForValue(sa.getResult(), node->src, sv, foldFlow(sa));
-}
-
-void FieldSource::visitMem(MemOp mem) {
-  for (auto r : mem.getResults())
-    makeNodeForValue(r, r, {}, foldFlow(r));
-}
-
-void FieldSource::visitInst(InstanceOp inst) {
-  for (auto r : inst.getResults())
-    makeNodeForValue(r, r, {}, foldFlow(r));
-}
-
-void FieldSource::visitInstChoice(InstanceChoiceOp inst) {
-  for (auto r : inst.getResults())
-    makeNodeForValue(r, r, {}, foldFlow(r));
-}
-
-const FieldSource::PathNode *FieldSource::nodeForValue(Value v) const {
+  // Check cache.
   auto ii = paths.find(v);
-  if (ii == paths.end())
-    return nullptr;
-  return &ii->second;
+  if (ii != paths.end())
+    return ii->second;
+  
+  Operation* op = v.getDefiningOp();
+  return TypeSwitch<Operation*, PathNode>(op)
+  .Case<SubfieldOp>([&](auto subf) {
```

Comment:
Should this be able to look through cast operations? I'd guess "yes" but am not sure.

Reply:
Casts are weird.  We've generally treated them as "reads" or rather "result of a cast is a source flow (rhs)".  In which case you leave pointer realm and get into value realm.